app/src/main/java/app/nexd/android/ui/auth/register/RegisterFragment.kt

```diff
@@ -46,42 +46,43 @@ class RegisterFragment : Fragment() {
         }
 
         button_register.setOnClickListener {
-            viewModel.register()
+            when (viewModel.progress.value) {
+                // we might come back using navigation and can't register again
+                is Registered -> findNavController().navigate(toRegisterDetailedFragment())
+                else -> viewModel.register()
+            }
         }
 
         viewModel.progress.observe(viewLifecycleOwner, Observer { progress ->
             progressBar.visibility = View.GONE
-            editText_first_name.isEnabled = true
-            editText_last_name.isEnabled = true
-            editText_email.isEnabled = true
-            editText_password.isEnabled = true
-            editText_password_confirm.isEnabled = true
+            enableInputFields()
 
             when (progress) {
                 is Idle -> { /* nothing to do here */
                 }
                 is Loading -> {
                     progressBar.visibility = View.VISIBLE
-                    editText_first_name.isEnabled = false
-                    editText_last_name.isEnabled = false
-                    editText_email.isEnabled = false
-                    editText_password.isEnabled = false
-                    editText_password_confirm.isEnabled = false
+                    disableInputFields()
                 }
                 is Error -> {
                     DefaultSnackbar(view, progress.message, Snackbar.LENGTH_SHORT)
                 }
                 is Finished -> {
-                    findNavController().navigate(
-                        toRegisterDetailedFragment()
-                    )
+                    disableInputFields()
+                    viewModel.progress.value = Registered
```

Comment:
Actually the navigation should not happen automatically on a certain progress state, since then it happens always and the user can never go back. The ViewModel should trigger a one time operation to navigate, the fragment should not need to know about for which progress state a navigation event should happen, only that it should navigate when triggered once. The additional state is just a workaround to avoid more refactoring.

Reply:
What i actually meant is that a Fragment should never change the progress state of the ViewModel directly. Progress abstracts inner state of the ViewModel and provides this way the view a possibility to adapt its views (e.g. by navigating or hiding/showing views). Especially in this case Fragment is changing the state of the ViewModel directly based on an other state of ViewModel (Finished -> Registered). 

The navigation is actually completely in the responsibility of the view layer. Theoretically the view could be refactored to use scrolling instead of navigation and the ViewModel could stay the same (by using the progress state). And if UI decides to navigate on a certain progress change (e.g. Finished), then it is ok. 

Also if you want a ViewModel to trigger some LiveData event once (without caching), maybe SingleLiveEvent class could be useful for you. 

In general I think that all these issues would not exist if we just not provide the "back" possibility here. In the end the user can't do anything on the first screen anyway after completing it.

Reply:
I think an easy solution would be to make the public liveData immutable.

Reply:
I also think a single event from the ViewModel is the way to go here.